### ISSUE TITLE
Aur search can't handle dashes in the search term

### ISSUE THREAD

**Issue by RubenKelevra** [fosskers/aura#869]
I'm currently using Aura 4.0.0, and I noticed that Aura can't handle dashes in the search term, if searching the AUR:

```
$ aura -Ss firefox-extension
extra/firefox-extension-mailvelope 5.1.2-1 (firefox-addons)
    Browser extension for OpenPGP encryption with Webmail
extra/firefox-extension-passff 1.17.0-1 (firefox-addons)
    zx2c4 pass manager addon for firefox

$ aura -As firefox-extension

```

Despite there being search results:

```
aura -As firefox | grep extension
aur/firefox-extension-video-download-helper 9.0.2.12-1 (3 | 0.00) 
    Adds a 'firefox'-symlink for 'firefox-developer-edition'. Also symlinks extensions, icon- and .desktop-files.
aur/firefox-extension-undo-close-tab-button 1.7-1 (1 | 0.00) 
    Browser extension that adds toolbar buttons for quick access to undo the closing of tabs
aur/firefox-extension-gooreplacer 3.13.0-2 (1 | 0.00) 
    A browser extension to modify HTTP requests
aur/firefox-extension-canvasblocker-bin 1.10.1-2 (1 | 0.09) 
aur/firefox-extension-facebook-container 2.3.9-1 (1 | 0.00) 
    Firefox extension that enables interacting with Ethereum Virtual Machine networks enabled websites
aur/extension-downloader-git v0.1.0.r7.gf683f75-1 (0 | 0.00) 
    Download browser extensions for Firefox and Chromium-based browsers
aur/firefox-esr-extension-download-with-aria2 4.10.0.2702-1 (0 | 0.00) 
aur/firefox-esr-extension-privacybadger 2024.6.14-1 (0 | 0.00) 
aur/firefox-extension-enhancer-for-youtube-bin 2.0.126-2 (0 | 0.00) 
aur/firefox-extension-overbitenx 0.9.2.1-1 (0 | 0.00) 
aur/firefox-extension-violentmonkey-bin 2.19.0-1 (0 | 0.00) 
aur/firefox-extension-return-youtube-dislike-bin 3.0.0.13-1 (0 | 0.00) 
```

**Comment by fosskers** [comment]
> Aura can't handle dashes in the search term

For that, search by each term individually.

```
> aura -As firefox extension --limit 5
aur/firefox-extension-arch-search 1.0.5-2 (18 | 0.02) 
    A set of Web Extensions that adds ArchLinux (bug tracker, forum, packages, wiki, AUR) as a search engine to the Firefox browser.
aur/firefox-extension-greasemonkey 4.11-1 (17 | 0.00) 
    Customize the way a web page displays or behaves, by using small bits of JavaScript.
aur/chrome-token-signing 1:1.1.5.547-1 (15 | 0.00) 
    Chrome and Firefox extension for signing with your eID on the web
aur/firefox-extension-keepassxc-browser 1.9.1.1-1 (12 | 0.00) 
    Official browser plugin for the KeePassXC password manager.
aur/firefox-extension-plasma-integration-bin 1.9.1-2 (7 | 0.12) 
    Multitask efficiently by controlling browser functions from the Plasma desktop.
```

>  aura -As firefox --reverse | tail -n 20

You can use `--limit` instead of a manual `tail`.

> I'd expect to see firefox-nightly, firefox-beta-bin, librewolf-bin before firefox-esr-bin

Confirmed. This isn't Aura's fault per se but something in the Faur. I'll look into it.

**Comment by RubenKelevra** [comment]
> For that, search by each term individually.

Errrr the issue here is not, that you can't do something different, the issue is, that it breaks the principle of the least astonishment. One should be able to search for a package name, without having to remove the dashes.

**Comment by fosskers** [comment]
As mentioned above, Aura 4 uses [its own metadata server](https://git.sr.ht/~fosskers/faur). From the `faur` README:

> Caveat: `by=desc` is term-based, not regex based. This is for performance reasons. So, `packages?names=aura&by=desc` will match on aura-bin but not on auralcap.

To vastly improve performance in the term-based search case, regex/substring-based searching is not possible. I realise this may be surprising and I apologise for that. The difference is simply due to the other package managers calling the AUR API directly. The Faur does other things that the AUR API cannot, but to enable those this sacrifice was made.

**Comment by fosskers** [comment]
The issue of `firefox-nightly` etc. not appearing in the search terms has been fixed on the Faur side. You should be able to see that in your current version of Aura.

**Comment by ultimatespirit** [comment]
> As mentioned above, Aura 4 uses [its own metadata server](https://git.sr.ht/~fosskers/faur). From the `faur` README:
> 
> > Caveat: `by=desc` is term-based, not regex based. This is for performance reasons. So, `packages?names=aura&by=desc` will match on aura-bin but not on auralcap.
> 
> To vastly improve performance in the term-based search case, regex/substring-based searching is not possible. I realise this may be surprising and I apologise for that. The difference is simply due to the other package managers calling the AUR API directly. The Faur does other things that the AUR API does not, but to enable those this sacrifice was made.

Can we get a fallback option then? It's for all intents and purposes a regression to be incapable of searching for a direct package name at all. At least let a search reach out to the normal AUR api which seems to be stronger here if the user asks for it.

**Comment by fosskers** [comment]
If you search for a direct package name, it will find it. More specifically, if you pass multiple terms, like:

```
aura -As nintendo switch
```
Then any package which contains both `nintendo` and `switch` in either its name, description, or keywords will be returned. Package names are split on `-` and `_` for individual terms. There is then a cleaning process that strips off various symbols and filters out terms that are too short, etc. What remains form "indexes" by which future lookups are very fast.

**Comment by RubenKelevra** [comment]
> Then any package which contains both `nintendo` and `switch` anywhere in its name, description, or keywords will be returned. Package names are split on `-` and `_` for individual terms.

Why not also add sections of multiple words to the index?

Like `firefox-nightly-i18n-xh` is adding these instead of just the individual words:

```
firefox
nightly
i18n
xh
firefox-nightly
firefox-nightly-i18n
firefox-nightly-i18n-xh
i18n-xh
nightly-i18n-xh
nightly-i18n
```

**Comment by fosskers** [comment]
One easy compromise is for Aura, on its end, to split the search input and send the split query to the Faur instead. So:
```
aura -As firefox-extension
```
would yield the same results as 
```
aura -As firefox extension
```

**Comment by fosskers** [comment]
As discussed, the above two searches now yield the same results.

**Comment by fosskers** [comment]
I've released `4.0.1`, I'm going to close this.
